[core] Emit an Event whenever a Grammar is AutoAssigned

confused-Techie opened the pull request:
This PR adds a new event that's emitted whenever the `GrammarRegistry` auto assigns a grammar to a file.

The event will contain the payload of:

```
{
  grammar
  buffer
}
```

This change is required for the feature I'm attempting to add to core that would allow auto-finding of packages to support unsupported file extensions.

But this event can be used to emit always, so that if there is some other potential use for this information other community packages can take advantage.

DeeDeeG reviewed (approved):
Seems fine to me!

We can "wait until there's a user for it", to follow the example of the Linux kernel project's conventions? Or merge it now ahead of its first user. Either way, I guess.

confused-Techie commented:
@DeeDeeG Thanks for the approval! I'm happy to wait for the first user of it, which I'm assuming you know is in #909

mauricioszabo commented:
Some questions on this:

First, is that I'm unsure when this event is going to be emitted. Suppose I open a Clojure file - will this event be emitted? And what if I open a `.ini` file, will it be emitted with something like "unknown format"?

Second is - usually, we treat event emitters as implementation details. Care to add an API `atom.grammars.onDidAutoAssignGrammar` and add documentation to it explaining the first question I had?

confused-Techie commented:
@mauricioszabo To my knowledge and testing this would only ever be triggered when being auto assigned to "Plain Text" as a fallback.

Meaning that if "Plain Text" is the right grammar, such as opening a `.txt` document, this won't be triggered. Then if a grammar is found for your language this would never be triggered. So yeah to my knowledge it only triggers in the event of us not being able to find any appropriate grammar at all for your currently open buffer.

DeeDeeG commented:
If anything, a tweak to the PR title might clear up the use-case for this.

confused-Techie commented:
@DeeDeeG In my testing this function does exactly as I described. But we can wait on this one for further testing to ensure that if we would like, since those tests were done a while ago and things are a bit foggy. But also the purpose of this PR is to interact with #909 so in my mind doesn't really have much urgency to be merged until I finish the feedback on that PR, which I've been lacking on doing so

savetheclocktower reviewed (changes requested):
I think we can consider this in isolation instead of treating it as linked to #909.

I would want two things added before 👍:

* an `onDidAutoAssignGrammar` function, as @mauricioszabo requested
* new specs to demonstrate this functionality works

Because whether #909 is ready or not, it's worth exposing this functionality to consumers in general.

confused-Techie commented:
Thanks to everyone's feedback, and thanks to @savetheclocktower for getting me to think of this one again.

I've now added specs, and determined I was initially wrong about when this would fire, and as such have also fixed it.

Would appreciate anyone that's able to take a look and see what we can do about this one!
